octodns/provider/ns1.py

```diff
@@ -193,22 +197,56 @@ def notifylists_list(self):
         return self._try(self._notifylists.list)
 
     def records_create(self, zone, domain, _type, **params):
-        return self._try(self._records.create, zone, domain, _type, **params)
+        cached = self._records_cache.setdefault(zone, {}) \
+            .setdefault(domain, {})
+        cached[_type] = self._try(self._records.create, zone, domain, _type,
+                                  **params)
+        return cached[_type]
 
     def records_delete(self, zone, domain, _type):
+        try:
+            # remove record's zone from cache
+            del self._zones_cache[zone]
+            # remove record from cache, after zone since we may not have
+            # fetched the record details
+            del self._records_cache[zone][domain][_type]
+        except KeyError:
+            # never mind if record is not found in cache
+            pass
         return self._try(self._records.delete, zone, domain, _type)
 
     def records_retrieve(self, zone, domain, _type):
-        return self._try(self._records.retrieve, zone, domain, _type)
+        cached = self._records_cache.setdefault(zone, {}) \
+            .setdefault(domain, {})
+        if _type not in cached:
+            cached[_type] = self._try(self._records.retrieve, zone, domain,
+                                      _type)
+        return cached[_type]
 
     def records_update(self, zone, domain, _type, **params):
-        return self._try(self._records.update, zone, domain, _type, **params)
+        cached = self._records_cache.setdefault(zone, {}) \
+            .setdefault(domain, {})
+        try:
+            # remove record's zone from cache
+            del self._zones_cache[zone]
+            # remove record from cache, after zone since we may not have
+            # fetched the record details
+            del cached[_type]
+        except KeyError:
+            # never mind if record is not found in cache
+            pass
```

Comment:
I feel like it might better to wrap each `del` in its own try-except in all methods. Current block assumes that if zone cache doesn't exist, record cache won't exist either. But `records_update` clears zone cache then re-populates record cache which invalidates the assumption. It is practically unlikely we'll run into this now, but I'd rather make it harder to make mistakes in cache management for future code changes. It'll probably be useful to have a clear_cache helper method. Let me take a stab at it.

Reply:
Decided to go a little fancy with decorators for a more centralized cache management.

Reply:
Came to the same realization this morning. Subsequent deletes of records in the same zone wouldn't work as the zone wouldn't exist and it wouldn't make it to the record. So either one first doesn't work. So definitely have to delete both always.